internal/controller/apiextensions/composite/environment_fetcher.go

```diff
@@ -185,3 +194,14 @@ func mergeMaps(a, b map[string]interface{}) map[string]interface{} {
 	}
 	return out
 }
+
+func validateEnvironment(req EnvironmentFetcherRequest, env *Environment) error {
+	if req.Revision == nil || req.Revision.Spec.Environment == nil || req.Revision.Spec.Environment.Schema == nil {
+		return nil
+	}
+	schema := &openapispec.Schema{}
+	if err := json.Unmarshal(req.Revision.Spec.Environment.Schema.JSONSchema.Raw, schema); err != nil {
+		return errors.Wrap(err, errParseEnvironmentSchema)
+	}
+	return openapivalidate.AgainstSchema(schema, env.Object, openapistrfmt.Default)
```

Comment:
how do we handle the situation when `env` is `nil`?

Reply:
`env` is never nil because `fetchEnvironment` always returns a value (if not an error).

Reply:
in that case, we could change the function signature to

```go
func validateEnvironment(req EnvironmentFetcherRequest, env Environment) error
```

Reply:
I would prefer to pass a pointer and save some performance by avoiding copying to much data around. In the standard usage `env` is never nil so it makes no difference. However, I added an additional check to set `env` to an empty object to prevent potential panics in the future.